src/components/tree-view/tree-view.tsx

```diff
@@ -57,58 +63,198 @@ export const TreeItem: React.SFC<TreeItemProps> = SBStateless(({ item, itemRende
 
 const TreeItemWrapper = observer(TreeItem);
 
+export class TreeStateMap {
+    stateMap: StateMap = new Map<TreeItemData, TreeItemState>();
+
+    getItemState(item: TreeItemData) {
+        const state = this.stateMap.get(item);
+        if (state) {
+            return state;
+        } else {
+            const newState = observable({ isSelected: false, isExpanded: false, isFocused: false });
+            this.stateMap.set(item, newState);
+            return newState;
+        }
+    }
+}
+
 @SBComponent(style) @observer
 export class TreeView extends React.Component<TreeViewProps, {}>{
-    static defaultProps = { itemRenderer: TreeItemWrapper, onSelectItem: () => {} };
+    static defaultProps: Partial<TreeViewProps> = { itemRenderer: TreeItemWrapper, onSelectItem: () => {}, onFocusItem: () => {} };
 
-    stateMap: StateMap = new Map<TreeItemData, TreeItemState>();
+    stateMap: TreeStateMap = new TreeStateMap();
+    parentsMap: ParentsMap = new Map<TreeItemData, TreeItemData | undefined>();
 
     constructor(props: TreeViewProps) {
         super(props);
-        this.initStateMap(props.dataSource);
+        this.initParentsMap(props.dataSource as TreeItemData[], undefined);
     }
 
-    initStateMap(data: Object[] = []) {
+    initParentsMap(data: TreeItemData[] = [], parent: TreeItemData | undefined) {
         data.forEach((item: TreeItemData) => {
-            this.stateMap.set(item, observable({ isSelected: false, isExpanded: false }));
-            this.initStateMap(item.children || []);
+            this.parentsMap.set(item, parent);
+            this.initParentsMap(item.children || [], item);
         });
     }
 
     componentDidMount() {
         autorun(() => {
             if (this.props.selectedItem) {
-                this.stateMap.get(this.props.selectedItem)!.isSelected = true;
+                action(() => this.stateMap.getItemState(this.props.selectedItem!).isSelected = true)();
```

Comment:
I think having methods for getting current selected / focused items state will make the code more readable